### PR TITLE
Enhance SM3 engine performance

### LINKED ISSUE #824: Enhance SM3 engine performance
`SM3Engine` should not copy a word from the input.
That would enhance the performance a bit.

### PR DESCRIPTION
**Pull request by icyclv**
`SM3Engine` should not copy a word from the input in some cases. This can improve performance by about 3%.

This PR will resolves #824 .

### CODE REVIEW

File: kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java [hunk 1]
```diff
@@ -104,7 +104,8 @@ public void update(byte message) {
         word[wordOffset++] = message;
 
         if (wordOffset >= word.length) {
-            processWord();
+            processWord(this.word,0);
```
Comment (johnshajiang) [inline kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java:107]: `processWord(word, 0)`

File: kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java [hunk 2]
```diff
@@ -122,20 +123,17 @@ public void update(byte[] message, int offset, int length) {
             while (consumed < length) {
                 word[wordOffset++] = message[offset + consumed++];
                 if (wordOffset >= word.length) {
-                    processWord();
+                    processWord(this.word,0);
```
Comment (johnshajiang) [inline kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java:126]: `processWord(word, 0)`

File: kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java [hunk 3]
```diff
@@ -168,13 +166,12 @@ public byte[] doFinal() {
...
         block[blockOffset]
-                = ((word[0] & 0xFF) << 24)
-                | ((word[1] & 0xFF) << 16)
-                | ((word[2] & 0xFF) << 8)
-                | ((word[3] & 0xFF));
-        wordOffset = 0;
+                = ((word[offset] & 0xFF) << 24)
+                | ((word[offset+1] & 0xFF) << 16)
+                | ((word[offset+2] & 0xFF) << 8)
+                | ((word[offset+3] & 0xFF));
```
Comment (johnshajiang) [inline kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java:174]:
```
offset + 1
offset + 2
offset + 3
```

**Review by johnshajiang — CHANGES_REQUESTED** [review]
`Copyright (C) 2022, 2023` --> `Copyright (C) 2022, 2024`

File: kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java [hunk 4]
```diff
@@ -104,7 +104,8 @@ public void update(byte message) {
         word[wordOffset++] = message;
 
         if (wordOffset >= word.length) {
-            processWord();
+            processWord(word, 0);
+            this.wordOffset = 0;
```
Comment (johnshajiang) [inline kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java:108]: I would have mentioned this line: `wordOffset = 0`

**Review by johnshajiang — CHANGES_REQUESTED** [review]

File: kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java (hunk 4 above)
Comment (icyclv) [reply kona-crypto/src/main/java/com/tencent/kona/crypto/provider/SM3Engine.java]: Sorry for the oversight. I rechecked it, and I guess this commit should be fine now.

**Review by johnshajiang — APPROVED** [review]
LGTM

**Pull request merged**